SDD 0025: Document Commodore component instantiation (#110)
projectsyn/documentation: docs/modules/SDDs/pages/0025-commodore-component-instantiation.adoc

= SDD 0025 - Commodore Component Instantiation

:sdd_author:    Simon Gerber
:sdd_owner:     Simon Gerber
:sdd_reviewers: Reviewer names
:sdd_date:      2020-10-23
:sdd_status:    draft
include::partial$meta-info-table.adoc[]

[NOTE]
.Summary
====
This SDD describes possible options to allow Commodore components to be instantiated multiple times per cluster.
It additionally defines the option we've chosen to implement.
====

== Motivation

Currently Commodore components don't support installing multiple instances of the software they manage.

> There are already a some components for which installing multiple instances is a common use case, such as the https://github.com/projectsyn/component-nfs-client-provisioner/[nfs-client-provisioner component].

Reviewer: ```suggestion
There are already some components for which installing multiple instances is a common use case, such as the https://github.com/projectsyn/component-nfs-client-provisioner/[nfs-client-provisioner component].
```

=== Goals

* Define how we want to allow Commodore components to be instantiated multiple times per cluster.

=== Non-Goals

* Define a solution for managing multiple instances of services for the customer on a Syn-enabled cluster.

== Design Proposal

Commodore will provide a well-defined structure for instantiating components and providing per-instance parameters.
Commodore will provide the instance's identifier ("name") to the component in a well-defined way.

> Component authors must explicitly mark their components as supporting instantiation.

Reviewer: That is a strange sentence.

```suggestion
Component authors must explicitly declare that their component is ready for being instantiated multiple times.
```

Well, not that happy with that wording either. Feel free to come up with something better.

Author: I took half of your suggestion and half of the original :)

Similar to Helm charts, the components themselves must make sure to not cause any naming collisions of objects

> This is required both for namespaces and non-namespaced resources.

Reviewer: ```suggestion
This is required both for namespaced and non-namespaced resources.
```

A pure form of this approach needs no changes in Commodore, but has some significant drawbacks from a usability perspective.
For this approach, component authors need to implement multi-instance support in each component individually.

This approach works well for infrastructure tooling for which a Kubernetes operator exists.
If an operator exists, the component can install the operator and associated CRDs, and can render instances as custom resources for the operator to provision.

The main drawback of the pure approach is that it may not be feasible to implement without some changes in Commodore for components which use Helm charts as their base, unless the Helm chart exposes a similar structure.
The reason for this is that Kapitan's Helm templating is expressed in the reclass inventory, and reclass isn't flexible enough to instantiate multiple copies of a part of the hierarchy based on another key in the inventory.

While it's ultimately left to the component authors, we assume that many components will expose the multi-instance nature of the component in `parameters.<component-name>`.

> Taking `nfs-client-provisioner` as our working example, the parameters structure might look roughly as reproduced below.

Reviewer 2: Another example:

```
applications:
  - nfs-client-provisioner
  - nfs-client-provisioner as nfs-server1
  - nfs-client-provisioner as nfs-server2

parameters:
  nfs_client_provisioner:
    namespace: syn-nfs-provisioner
    host: nfs.example.com
    path: /nfs
  nfs_server1:
    host: nfs1.example.com
  nfs_server2:
    host: nfs2.example.com
```

1. Using the `as` keyword in an application (component include) string instantiates the component with the defined alias.
2. Commodore will merge the dict `$alias` with `nfs_client_provisioner` to provide/override defaults.
3. Inside the component a helper function wraps `kap.inventory().parameters.$alias` to provide the correct values.

If no alias is provided, the component is imported with it's default name.

Author: > 3\. Inside the component a helper function wraps `kap.inventory().parameters.$alias` to provide the correct values.

This is not even necessary, as we can simply merge the alias config values into `nfs_client_provisioner` in the target for the aliased instance:

```
# inventory/targets/nfs-server1.yml
classes:
- [ ... defaults ... ]
- global.commodore
- components.nfs-client-provisioner

parameters:
  kapitan:
    vars:
      target: nfs-server1.yml
  nfs_client_provisioner: ${nfs_server1}

Author: Also note that this suggestion does not work for the no Commodore support approach, and the original example is not what will be implemented. I'll update the "Design Proposal" section, with this example approach and my variant of it.
